tests/python/model/test_ratex_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,134 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Test ZeRO-1 implementation using FSDP interface."""
+import pytest
+
+import torch
+import torch.nn as nn
+from torchvision import datasets, transforms
+
+from raf.testing import get_dist_comm_info
+from raf import distributed as dist
+
+import ratex
+import ratex.lazy_tensor_core.core.lazy_model as lm
+from ratex.core.distributed.ratex_fully_sharded_data_parallel import RatexFullyShardedDataParallel
+from ratex.optimizer import SGD, Adam
+
+import numpy as np
+
+
+class SingleLayerLogistics(nn.Module):
+    def __init__(self, input_shape=28, num_classes=10):
+        super(SingleLayerLogistics, self).__init__()
+        self.log_softmax = nn.LogSoftmax(dim=-1)
+        self.linear = nn.Linear(784, num_classes)
+
+    def forward(self, x):
+        out = torch.flatten(x, 1)
+        out = self.linear(out)
+        out = torch.relu(out)
+        out = self.log_softmax(out)
+        return out
+
+
+def train(
+    device,
+    model,
+    optimizer,
+    optimizer_config,
+    image_datasets,
+    fsdp=False,
+    num_epochs=10,
+    dtype=torch.float32,
+):
+    dataloaders = {
+        x: torch.utils.data.DataLoader(
+            image_datasets[x], batch_size=1, shuffle=False, num_workers=1
+        )
+        for x in ["train", "val"]
+    }
+    dataset_sizes = {x: len(image_datasets[x]) for x in ["train", "val"]}
+    model = model.to(device, dtype=dtype)
+    if fsdp:
+        model = RatexFullyShardedDataParallel(model, optimizer, optimizer_config)
+        optimizer = model
+        model.train()
+    else:
+        model.train()
+        optimizer = optimizer(model.parameters(), **optimizer_config)
+    model = ratex.jit.script(model)
+
+    for epoch in range(num_epochs):
+        running_loss = 0.0
+
+        # Iterate over data.
+        for inputs, labels in dataloaders["train"]:
+            inputs = inputs.to(device)
+            inputs.requires_grad = True
+            labels_one_hot = torch.tensor(np.eye(10, dtype=np.float32)[labels])
+            labels_one_hot = labels_one_hot.to(device)  # One-hot
+            optimizer.zero_grad()
+            outputs = model(inputs)
+            loss = -torch.sum(outputs * labels_one_hot) / inputs.size(0)
+            loss.backward()
+            optimizer.step()
+            lm.mark_step()
+            running_loss += loss.item() * inputs.size(0)
```

Comment:
Although it's not a big deal in a unit test, it's better to follow the best practice of lazy tensor. Specifically, in this line you materialize the loss by `.item()`. It creates a barrier at this point and prevents the overlapping between execution and graph tracing. See https://github.com/awslabs/ratex/blob/main/docs/overlap_trace_and_computation.md for details.

Reply:
I changed it so that the loss is not materialized until the end of each epoch.